System/FilePath/Internal.hs

```diff
@@ -984,9 +985,12 @@ normaliseDrive :: FILEPATH -> FILEPATH
 normaliseDrive bs
   | null bs = mempty
   | isPosix = pack [pathSeparator]
-  | otherwise = if isJust $ readDriveLetter x2
-         then map toUpper x2
-         else x2
+  | Just (drv, _) <- readDriveLetter x2
+                  = case unpack drv of
+                      (x:_:[]) -> pack [toUpper x, _colon]
+                      (x:_) -> pack [toUpper x, _colon, pathSeparator]
```

Comment:
Is it expected to have anything else than `:` as the second element of the list?

Reply:
No, but `readDriveLetter` is maintaining this invariant:

```
readDriveLetter :: STRING -> Maybe (FILEPATH, FILEPATH)
readDriveLetter bs = case unpack bs of
  (x:c:y:xs)
    | c == _colon && isLetter x && isPathSeparator y
    -> Just $ addSlash (pack [x,_colon]) (pack (y:xs))
  (x:c:xs)
    | c == _colon && isLetter x
    -> Just (pack [x,_colon], pack xs)
  _ -> Nothing
```

Reply:
We could inline this, but...